src/common/threading_utils.h

```diff
@@ -246,6 +246,43 @@ inline int32_t OmpGetNumThreads(int32_t n_threads) {
   n_threads = std::max(n_threads, 1);
   return n_threads;
 }
+
+
+/*!
+ * \brief A C-style array with in-stack allocation. As long as the array is smaller than
+ * MaxStackSize, it will be allocated inside the stack. Otherwise, it will be
+ * heap-allocated.
+ */
+template <typename T, size_t MaxStackSize>
+class MemStackAllocator {
+ public:
+  explicit MemStackAllocator(size_t required_size) : required_size_(required_size) {
+    if (MaxStackSize >= required_size_) {
+      ptr_ = stack_mem_;
+    } else {
+      ptr_ = reinterpret_cast<T*>(malloc(required_size_ * sizeof(T)));
+    }
+    if (!ptr_) {
+      throw std::bad_alloc{};
+    }
+  }
+
+  ~MemStackAllocator() {
+    if (required_size_ > MaxStackSize) {
+      free(ptr_);
+    }
+  }
+  T& operator[](size_t i) { return ptr_[i]; }
+  T const& operator[](size_t i) const { return ptr_[i]; }
+
+  // FIXME(jiamingy): Remove this once we merge partitioner cleanup for hist.
+  auto Get() { return ptr_; }
```

Comment:
Can we remove this now? `GHistIndexMatrix` is no longer using `Get()` method.

Reply:
@hcho3 It's still used by the hist updater.